### PR TITLE
[dynamo] Represent all cells as `NewCellVariable`

### LINKED ISSUE #137123: Investigate making `prune_dead_locals` more aggressive by tracing liveness from variables in `SideEffects`
This serves as a tracking issue to follow up conversation from another PR: https://github.com/pytorch/pytorch/pull/136891#discussion_r1783356678.

cc @ezyang @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @amjames @rec

### PR DESCRIPTION
**Pull request by StrongerXi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #140155
* #140154
* __->__ #140153
* #140435
* #140436
* #140152
* #140330

In addition to `NewCellVariable`, Dynamo has 3 ways of modeling cell objects:
1. For cells captured and created by the root frame, represent them as
   their contents in `root_tx.symbolic_locals`, which `LOAD_DEREF` and
   `STORE_DEREF` update directly, without going through `SideEffects`.
2. `ClosureVariable`: this is created when cells from (1) are captured
   by a newly created function Dynamo is about to inline. It's a handle
   with a name that redirects `LOAD_DEREF` and `STORE_DEREF` back (1),
   to make `root_tx.symbolic_locals` up-to-date.
3. For cells that are captured by both the root frame and some
   pre-existing function Dynamo is about to inline, represent those
   cells as contents, and do not allow writes to them.

Note that (2) and (3) are mainly to conform with (1) -- to make sure
Dynamo has a consistent modeling of cells for the same cell objects.

In this patch, we represent all of these cells as `NewCellVariable`. The
main new code paths introduced are:
- using `NewCellVariable` to model cell objects created by the root
  frame (the cells are passed in as input to `InstructionTranslator`),
  this is what allows us to get rid of all 3 legacy paths above.
- adding a new `AutoDerefLocalSource` to deal with the python-code
  level (guards) and bytecode level (codegen) auto-dereferencing
  behavior, when accessing pre-existing python cells. This also
  involves a tiny update to guard manager generation.
- plumbing some extra info into `LocalSource` and `CellVariable` so that
  we can still emit `LOAD_DEREF`, `STORE_DEREF`, `LOAD_CLOSURE` (instead
  of `make_cell`, `cell_contents` attribute access, and `LOAD_FAST`),
  which is important for readability, performance, and some
  assumptions `bytecode_transformation.py` makes.

As a result, this patch removes a lot of the now-dead code paths and
TODOs. Notably, it significantly simplified the `prune_dead_locals`
function, which was duplicating a lot of the logic from
`prune_dead_object_new`; this conveniently closes #137123.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

**Comment by StrongerXi** [comment]
Tests are failing, here's what I've found so far:
1. Python 3.9 and 3.10 fail because the frame object we pass to Dynamo's eval frame callback doesn't have `f_func: FunctionType` (this patch uses it to retrieve the cells captured by the root frame).
2. This patch exposed a bug in Dynamo's handling of `out=` keyword for torch operators like `sort` -- the Python semantics is in-place mutation on the underlying tensor object, but in Dynamo, we create a new `TensorVariable` (through `wrap_fx_proxy`) and tries to replace instances of the old `TensorVariable` in `symbolic_locals` with this new one. The replacement never accounted for variables that are not in `symbolic_locals`, but reachable from it (e.g., within a `TupleVariable`).
3. A `torch._dynamo.exc.Unsupported: reconstruct: NewCellVariable()` failure that's 3.11 specific. I have some idea and will look more.

**Comment by StrongerXi** [comment]
I'll try to create separate issues for (2) and (3) above, and fix them. (2) feels a little annoying.

**Comment by StrongerXi** [comment]
Rebase.

File: torch/_dynamo/source.py [hunk 1]
```diff
@@ -224,6 +233,51 @@ def name(self):
         return f"{self.base.name()}.{self.member}"
 
 
+@dataclasses.dataclass(frozen=True)
+class AutoDerefLocalSource(ChainedSource):
```
Comment (StrongerXi) [inline torch/_dynamo/source.py:237]: This is mainly to make root-frame cells (the ones Python generates `LOAD_DEREF` and `STORE_DEREF` for) play well with guards. I think we can remove it if we make our `f_locals` (or whatever it becomes) contain cell objects without dereferencing them, which might be do-able as part of https://github.com/pytorch/pytorch/pull/140063#discussion_r1835483086? @jansel @williamwen42

File: torch/_dynamo/guards.py [hunk 2]
```diff
@@ -884,6 +886,14 @@ def get_guard_manager_from_source(self, source):
...
                 guard_manager_enum=guard_manager_enum,
             )
+        elif istype(source, AutoDerefLocalSource):
+            # Guard checks run on f_locals, in which the python level
+            # auto-dereferenced cell objects are also dereferenced (e.g., rather
+            # than `f_locals` being `{ 'cell' : <cell object of int> }`, it'll
+            # be `{ 'cell' : <int> }`. So the guard manager is the same as the
+            # base guard manager.
+            assert isinstance(base_guard_manager, GuardManager)  # tame mypy
+            out = base_guard_manager
```
Comment (StrongerXi) [inline torch/_dynamo/guards.py:896]: See https://github.com/pytorch/pytorch/pull/140153/files#r1840220615

File: test/dynamo/test_higher_order_ops.py [hunk 3]
```diff
@@ -971,25 +971,25 @@ def k(x):
...
-            getitem: "f32[s0, s2]" = wrap[0];  wrap = None
+            wrap = torch.ops.higher_order.wrap(wrap_body_0, s2, s0, l_x_, s1, l_y_);  wrap_body_0 = s2 = s0 = l_x_ = s1 = l_y_ = None
+            getitem: "f32[s2, s1]" = wrap[0];  wrap = None
             return (getitem,)
 
         class wrap_body_0(torch.nn.Module):
-            def forward(self, s0: "Sym(s0)", s1: "Sym(s1)", l_x_: "f32[s0, s1]", s2: "Sym(s2)", l_y_: "f32[s1, s2]"):
-                matmul: "f32[s0, s2]" = l_x_ @ l_y_;  l_x_ = l_y_ = None
+            def forward(self, s2: "Sym(s2)", s0: "Sym(s0)", l_x_: "f32[s2, s0]", s1: "Sym(s1)", l_y_: "f32[s0, s1]"):
+                matmul: "f32[s2, s1]" = l_x_ @ l_y_;  l_x_ = l_y_ = None
```
Comment (StrongerXi) [inline test/dynamo/test_higher_order_ops.py:992]: @ydwu4 does this change matter?

**Review by jansel — APPROVED** [review]

File: test/dynamo/test_higher_order_ops.py (hunk 3 above)
Comment (ydwu4) [reply test/dynamo/test_higher_order_ops.py:992]: The order doesn't matter for this hop as long as it's deterministic.

**Pull request closed without merging**